src/pymor/algorithms/projection.py

```diff
@@ -198,10 +197,32 @@ def action_LincombOperator(self, op):
     def action_SelectionOperator(self, op):
         return self.replace_children(op)
 
-    @match_class(OperatorInterface)
-    def action_generic_projection(self, op):
-        op.logger.warning('Using inefficient generic projection operator')
-        return ProjectedOperator(op, self.range_basis, self.source_basis, self.product)
+    @match_class(BlockOperatorBase)
+    def action_BlockOperatorBase(self, op):
+        if op.blocked_range:
+            if self.range_basis is not None:
+                range_basis = self.range_basis._blocks
+            else:
+                range_basis = [None] * len(op.range.subspaces)
+        else:
+            range_basis = [self.range_basis]
+        if op.blocked_source:
+            if self.source_basis is not None:
+                source_basis = self.source_basis._blocks
+            else:
+                source_basis = [None] * len(op.source.subspaces)
+        else:
+            source_basis = [self.source_basis]
+
+        projected_ops = np.array([[project(op.blocks[i, j], rb, sb)
+                                   for j, sb in enumerate(source_basis)]
+                                  for i, rb in enumerate(range_basis)])
+        if self.range_basis is None and op.blocked_range:
+            return BlockColumnOperator(np.sum(projected_ops, axis=1))
+        elif self.source_basis is None and op.blocked_source:
+            return BlockRowOperator(np.sum(projected_ops, axis=0))
+        else:
+            return np.sum(projected_ops)
```

Comment:
Is it guaranteed that `self.range_basis` and `self.source_basis` are not both `None`?

Reply:
See #791.